From b565ad053c5653f63a8137b835809857d987fc2f Mon Sep 17 00:00:00 2001
From: Dave Davenport <qball@blame.services>
Date: Tue, 12 Apr 2022 16:09:56 +0200
Subject: [PATCH] [Helper] Add XDG_DATA_DIRS to theme search path.

Issue: #1617
---
 source/helper.c | 25 +++++++++++++++++++++----
 1 file changed, 21 insertions(+), 4 deletions(-)

diff --git a/source/helper.c b/source/helper.c
index 559fdbd8..3dd7c3a8 100644
--- a/source/helper.c
+++ b/source/helper.c
@@ -1062,7 +1062,7 @@ char *helper_get_theme_path(const char *file, const char *ext) {
   const char *cpath = g_get_user_config_dir();
   if (cpath) {
     char *themep = g_build_filename(cpath, "rofi", "themes", filename, NULL);
-    g_debug("Opening theme, testing: %s\n", themep);
+    g_debug("Opening theme, testing: %s", themep);
     if (themep && g_file_test(themep, G_FILE_TEST_EXISTS)) {
       g_free(filename);
       return themep;
@@ -1072,7 +1072,7 @@ char *helper_get_theme_path(const char *file, const char *ext) {
   // Check config directory.
   if (cpath) {
     char *themep = g_build_filename(cpath, "rofi", filename, NULL);
-    g_debug("Opening theme, testing: %s\n", themep);
+    g_debug("Opening theme, testing: %s", themep);
     if (g_file_test(themep, G_FILE_TEST_EXISTS)) {
       g_free(filename);
       return themep;
@@ -1083,8 +1083,8 @@ char *helper_get_theme_path(const char *file, const char *ext) {
   if (datadir) {
     char *theme_path =
         g_build_filename(datadir, "rofi", "themes", filename, NULL);
-    g_debug("Opening theme, testing: %s\n", theme_path);
     if (theme_path) {
+      g_debug("Opening theme, testing: %s", theme_path);
       if (g_file_test(theme_path, G_FILE_TEST_EXISTS)) {
         g_free(filename);
         return theme_path;
@@ -1093,9 +1093,26 @@ char *helper_get_theme_path(const char *file, const char *ext) {
     }
   }
 
+  const gchar * const * system_data_dirs = g_get_system_data_dirs ();
+  if ( system_data_dirs ) {
+    for ( uint_fast32_t i = 0; system_data_dirs[i] != NULL; i++ ){
+      const char * const datadir = system_data_dirs[i];
+      g_debug("Opening theme directory: %s", datadir );
+      char *theme_path = g_build_filename(datadir, "rofi", "themes", filename, NULL);
+      if (theme_path) {
+        g_debug("Opening theme, testing: %s", theme_path);
+        if (g_file_test(theme_path, G_FILE_TEST_EXISTS)) {
+          g_free(filename);
+          return theme_path;
+        }
+        g_free(theme_path);
+      }
+    }
+  }
+
   char *theme_path = g_build_filename(THEME_DIR, filename, NULL);
   if (theme_path) {
-    g_debug("Opening theme, testing: %s\n", theme_path);
+    g_debug("Opening theme, testing: %s", theme_path);
     if (g_file_test(theme_path, G_FILE_TEST_EXISTS)) {
       g_free(filename);
       return theme_path;
-- 
2.33.1

